htdocs/commande/class/commande.class.php

```diff
@@ -596,6 +596,14 @@ public function valid($user, $idwarehouse = 0, $notrigger = 0)
 
 		if (!$error) {
 			$this->db->commit();
+			if (!$error && !$notrigger) {
+				// Call trigger
+				$result = $this->call_trigger('ORDER_POST_VALIDATE', $user);
```

Comment:
The goal of a trigger is to be inside a transaction so it can be canceled by the rollback so no trigger must be after commit()

Reply:
> The goal of a trigger is to be inside a transaction so it can be canceled by the rollback, so, no trigger must be after commit(). Never. 

ok. What if you want to be "notified", informed, messaged, ... about an after post transaction fact? Sort of like "heads up, this thing happened, deal with it"

Sometimes there is a value in getting a post happening "call to action". In this case it can be emailing the customer with new information about their purchase status. And it certainly before emailing it could be to regenerate the document.

> You can retrieve the new ref into property this->newref Generation of links that are using the old ref should be instead enhanced to user ->newref instead of ->ref when it is set. Can you work on a fix going in this direction ?

Perhaps, I'll have to look at it. Initially I don't really like that situation of just using different variables, and I don't think it makes good sense to move the post trigger to JUST before the commit, because it makes little sense to email the customer "accepted" when some other thing that listens to this trigger can cancel and then everything has been rolled back.


So how would you bring about a post transaction "trigger"/information/... ? (It could be named something like: Alleviate, Assauge, Commence, Debrief, Decommision, Mitigate, Mollify, Palliate, Placate, ...)

Would you make a whole duplication of the trigger system? but only for things that happened after a transaction has been committed?

Reply:
I am not testing in latest development commit, it is a tuxgasy v19 container, but here's my comments on that

> You can retrieve the new ref into property this->newref Generation of links that are using the old ref should be instead enhanced to user ->newref instead of ->ref when it is set. Can you work on a fix going in this direction ?

I just looked at what substitution variables is available when manually sending an order email, and here `__NEWREF__` is not available, only `__REF__` so I'll have to need 2 different email templates? one for manual sending and one for automatic notifications?

I also looked at this order using the API, and here the json does not contain any `newref`, only `ref` - and `"oldref": null,`


So those links that you say have to be updated to use `->newref instead of ->ref` - can I trust that ->newref is always available? What about substitution variables in email templates, will I have to use different variables here too depending if I'm sending manually or sending using the automatic notification system?


Sorry sir, but I don't really like this "solution" either. I wish that the only difference between manually send emails and automatic send emails would be the sender, I don't wish to have multiple templates or to use different variables in the json depending on it's status if it is draft or validated - or has been validated.

Also, there's the whole issue of potentially having 2 or more different subsystems that listens to the ORDER_VALIDATE trigger, does something based on that - like reserving a product/resource for something or emails the customer, ... and then the last thing that listens to the ORDER_VALIDATE trigger says no, because it is my understanding that they can say no and interrupt the action about to happen. Then what? automatic notifications already send the email.

Reply:
> What if you want to be "notified", informed, messaged, ... about an after post transaction fact? Sort of like "heads up, this thing happened, deal with it"

The trigger ORDER_VALIDATE is already executed at the end of the transaction. It is the last business action that has effect into database. So you should be able to use it for post action. All actions after are technical change that should not be modified by external code, like updating cache table.

the ->newref exists during the life of the trigger. So you can get it only inside code executed by a trigger.
When using an api, you will get only ref that is final ref.

You don't need several template. A template should use __REF__ . If the __REF__ does not contains the new ref when making the substitution, it is was i said when i say "Generation of links that are using the old ref should be instead enhanced to use ->newref instead of ->ref when it is set." 
It means there is something missing in the existing code in the core that make the substitution of __REF__. The current code does not take into account the ->newref field when it should.

Reply:
Hi

Do you have 2 code examples? One where the code uses ->newref and one where it does not use ->newref but it should?

> The ->newref exists on object during the life of the trigger. So you can get it only inside code executed by a triggers.

How does the code know if it is being executed inside a trigger?  Could you provide an example? then I can see if I can understand it.


> You don't need several template. A template should use `__REF__` (You should never have `__NEWREF__` inside a template, this key has no sense as a template is a user data and old ref / new ref is something managed on developer side). If the `__REF__` does not contains the new ref when making the substitution of this key, it is was i said when i say "Generation of links that are using the old ref should be instead enhanced to use ->newref instead of ->ref when it is set.". I mean this: There is something missing in the existing code in the core that make the substitution of `__REF__`. The current code does not take into account the ->newref field when it should.

If ->newref is supposed to be hidden, how come it showed up in the list of substitution variables in the emails I attached pictures of above?

Reply:
Where the code make the substitution __REF__ with ->ref, the code must instead do __REF__ replaced with ->newref (if ->newref is set).

Reply:
> Where the code make the substitution `__REF__` with ->ref, the code must instead do `__REF__` replaced with ->newref (if ->newref is set).

yeah? but where? I'm leaning towards closing this pull request and letting someone else fix the ref/newref quest.